working/333 - shared memory multithreading/shared_native_memory.md

```diff
@@ -0,0 +1,746 @@
+code is executed in such a way it can only access static state which is shared
+between isolates (`@pragma('vm:shared')`) and attempts to access isolated state
+will cause `FieldAccessError` to be thrown.
+
+```dart
+  /// Constructs a [NativeCallable] that can be invoked from any thread.
+  ///
+  /// When the native code invokes the function [nativeFunction], 
+  /// the [callback] will be executed within the isolate group
+  /// of the [Isolate] which originally constructed the callable. 
+  /// Specifically, this means that an attempt to access any 
+  /// static or global field which is not shared between 
+  /// isolates in a group will result in a [FieldAccessError].
+  ///
+  /// If an exception is thrown by the [callback], the 
+  /// native function will return the `exceptionalReturn`, 
+  /// which must be assignable to the return type of 
+  /// the [callback].
+  ///
+  /// [callback] and [exceptionalReturn] must be 
+  /// _trivially shareable_.
+  ///
+  /// This callback must be [close]d when it is no longer 
+  /// needed. An [Isolate] that created the callback will 
+  /// be kept alive until [close] is called.
+  ///
+  /// After [NativeCallable.close] is called, invoking 
+  /// the [nativeFunction] from native code will cause 
+  /// undefined behavior.
+  factory NativeCallable.isolateGroupBound(
+    @DartRepresentationOf("T") Function callback, {
+    Object? exceptionalReturn,
+  }) {
+    throw UnsupportedError("NativeCallable cannot be constructed dynamically.");
+  }
+```
+
+#### Core library API behavior 
+
+**TODO:** give exhaustive list of APIs that are expected to work outside of a
+concrete isolate. We expect all synchronous `dart:*` APis simply work. APIs
+which require message loop and microtask queue (e.g. async IO, timers,
+`Future` and `Stream` APIs) should throw easy to understand errors.
+
+### Additional Isolate APIs
+
+We should facilitate synchronously entering `Isolate` when necessary. 
+
+```dart
+class Isolate {
+  // Execute the given function in the context of the given isolate.
+  // 
+  // [f] must be trivially shareable. Result
+  // returned by [f] must be trivially shareable.
+  R runSync<R>(R Function() f);
+}
+```
+
+Note that `runSync` can only enter an `Isolate` when it is not used by
+another thread.
+
+**TODO**: Furthermore we might want to facilitate integration with third-party
+event-loops: e.g. allow to create isolate without scheduling its event loop on
+our own thread pool and provide equivalents of `Dart_SetMessageNotifyCallback`
+and `Dart_HandleMessage`. Though maybe we should not bundle this all together
+into one update.
+
+### Scoped thread local values
+
+```dart
+@pragma('vm:deeply-immutable')
+final class ScopedThreadLocal<T> {
+  /// Creates scoped thread local value with the given [initializer] function.
+  ///
+  /// [initializer] must be trivially shareable.
+  external factory ScopedThreadLocal([T Function()? initializer]);
```

Comment:
In the sample use below at line 344 `List<Object>` is not trivially shareable, right?

Reply:
There is no value of type `List<Object>` - the closure (`initializer`) is trivially shareable because it does not capture anything.